flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -824,11 +824,14 @@ public ProcessingTimeService getProcessingTimeService() {
 	 * FAILED, and, if the invokable code is running, starts an asynchronous thread
 	 * that aborts that code.
 	 *
-	 * <p>This method never blocks.</p>
+	 * <p>This method never blocks.
 	 */
 	@Override
 	public void handleAsyncException(String message, Throwable exception) {
-		getEnvironment().failExternally(exception);
+		if (isRunning) {
```

Comment:
I doubt that this is working fix, it will only make the problem less likely to occur. This method can be called asynchronously to Threads that manipulate `isRunning`, which means that the stream task can leave running status after the condition was checked as true, but before `failExternally(...)` went through.